hs/src/Reach/Eval.hs

```diff
@@ -1686,10 +1692,24 @@ evalApply ctxt at sco st rator rands =
         evalApplyVals ctxt at sco st_rands rator randsvs
       return $ SLRes (rlifts <> alifts) st_res r
 
+getKwdOrPrim :: Ord k => k -> M.Map k SLSSVal -> Maybe SLSSVal
+getKwdOrPrim ident env =
+  case M.lookup ident env of
+    Just s@(SLSSVal _ _ (SLV_Kwd _))  -> Just s
+    Just s@(SLSSVal _ _ (SLV_Prim _)) -> Just s
+    _ -> Nothing
+
 evalPropertyName :: SLCtxt s -> SrcLoc -> SLScope -> SLState -> JSPropertyName -> SLComp s (SecurityLevel, String)
 evalPropertyName ctxt at sco st pn =
   case pn of
-    JSPropertyIdent _ s -> k_res $ public $ s
+    JSPropertyIdent an s ->
+      -- Do not allow keywords or primitives to be used as property names
+      case getKwdOrPrim s base_env of
+        Just s' -> expect_throw at_n $ Err_Shadowed s s' dummy_at
+        _       -> k_res $ public s
+      where
+        at_n = srcloc_jsa "field" an at
+        dummy_at = SLSSVal at Public $ SLV_Null at ""
```

Comment:
The error message only looks at the position when reporting so I just used `null`, not sure if there's a better `SLVal` to represent an Object field?